File: packages/vm/wasmlib/contract.go
package wasmlib

type ScFuncCallContext interface {
	CanCallFunc()
}

type ScViewCallContext interface {
	CanCallView()
}

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

type ScView struct {
	hContract ScHname
	hFunction ScHname
	paramsID  *int32
	resultsID *int32
}

func NewScView(hContract, hFunction ScHname) *ScView {
	return &ScView{hContract, hFunction, nil, nil}
}

func (v *ScView) SetPtrs(paramsID, resultsID *int32) {
	v.paramsID = paramsID
	v.resultsID = resultsID
	if paramsID != nil {
		*paramsID = NewScMutableMap().MapID()
	}
}

func (v *ScView) Call() {
	v.call(0)
}

func (v *ScView) call(transferID int32) {
	encode := NewBytesEncoder()
	encode.Hname(v.hContract)
	encode.Hname(v.hFunction)
	encode.Int32(paramsID(v.paramsID))
	encode.Int32(transferID)
	Root.GetBytes(KeyCall).SetValue(encode.Data())
	if v.resultsID != nil {
		*v.resultsID = GetObjectID(OBJ_ID_ROOT, KeyReturn, TYPE_MAP)
	}
}

func (v *ScView) OfContract(hContract ScHname) *ScView {
	v.hContract = hContract
	return v
}

func paramsID(id *int32) int32 {
	if id == nil {
		return 0
	}
	return *id
}

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

type ScInitFunc struct {
	ScView
	keys       []Key
	indexes    []Key32
	oldIndexes []Key32
	host       ScHost
}

func NewScInitFunc(hContract, hFunction ScHname, ctx ScFuncCallContext, keys []Key, indexes []Key32) *ScInitFunc {
	f := &ScInitFunc{}
	f.hContract = hContract
	f.hFunction = hFunction
	if ctx == nil {
		f.keys = keys
		f.oldIndexes = append(f.oldIndexes, indexes...)
		f.indexes = indexes
		for i := 0; i < len(indexes); i++ {
			indexes[i] = Key32(i)
		}
		f.host = ConnectHost(NewInitHost())
	}
	return f
}

func (f *ScInitFunc) Call() {
	Panic("cannot call init")
}

func (f *ScInitFunc) OfContract(hContract ScHname) *ScInitFunc {
	f.hContract = hContract
	return f
}

func (f *ScInitFunc) Params() []interface{} {
	if f.keys == nil {
		Panic("cannot call params")
	}

	var params []interface{}
	for k, v := range host.(*InitHost).params {
		params = append(params, string(f.keys[k]))
		params = append(params, v)
	}
	copy(f.indexes, f.oldIndexes)
	ConnectHost(f.host)
	return params
}

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

type ScFunc struct {
	ScView
	delay      int32
	transferID int32
}

func NewScFunc(hContract, hFunction ScHname) *ScFunc {
	return &ScFunc{ScView{hContract, hFunction, nil, nil}, 0, 0}
}

func (f *ScFunc) Call() {
	if f.delay != 0 {
		Panic("cannot delay a call")
	}
	f.call(f.transferID)
}

func (f *ScFunc) Delay(seconds int32) *ScFunc {
	f.delay = seconds
	return f
}

func (f *ScFunc) OfContract(hContract ScHname) *ScFunc {
	f.hContract = hContract
	return f
}

func (f *ScFunc) Post() {
	f.PostToChain(Root.GetChainID(KeyChainID).Value())
}

func (f *ScFunc) PostToChain(chainID ScChainID) {
	if f.transferID == 0 {
		Panic("transfer is required for post")
	}
	encode := NewBytesEncoder()
	encode.ChainID(chainID)
	encode.Hname(f.hContract)
	encode.Hname(f.hFunction)
	encode.Int32(paramsID(f.paramsID))
	encode.Int32(f.transferID)
	encode.Int32(f.delay)
	Root.GetBytes(KeyPost).SetValue(encode.Data())
	if f.resultsID != nil {
		*f.resultsID = GetObjectID(OBJ_ID_ROOT, KeyReturn, TYPE_MAP)
	}
}

func (f *ScFunc) Transfer(transfer ScTransfers) *ScFunc {
	f.transferID = transfer.transfers.MapID()
	return f
}

func (f *ScFunc) TransferIotas(amount int64) *ScFunc {
	return f.Transfer(NewScTransferIotas(amount))
}

File: contracts/rust/wasmlib/src/contract.rs
// Copyright 2020 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

// base contract objects

use std::ptr;

use crate::bytes::*;
use crate::context::*;
use crate::hashtypes::*;
use crate::host::*;
use crate::keys::*;
use crate::mutable::*;

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

#[derive(Clone, Copy)]
pub struct ScView {
    h_contract: ScHname,
    h_function: ScHname,
    params_id: *mut i32,
    results_id: *mut i32,
}

impl ScView {
    pub fn zero() -> ScView {
        ScView {
            h_contract: ScHname(0),
            h_function: ScHname(0),
            params_id: ptr::null_mut(),
            results_id: ptr::null_mut(),
        }
    }

    pub fn new(h_contract: ScHname, h_function: ScHname, params_id: *mut i32, results_id: *mut i32) -> ScView {
        unsafe {
            if params_id != ptr::null_mut() {
                *params_id = ScMutableMap::new().map_id();
            }
        }

        ScView {
            h_contract: h_contract,
            h_function: h_function,
            params_id: params_id,
            results_id: results_id,
        }
    }

    pub fn call(&self) {
        self.call_with_transfer(0);
    }

    fn call_with_transfer(&self, transfer_id: i32) {
        let mut encode = BytesEncoder::new();
        encode.hname(&self.h_contract);
        encode.hname(&self.h_function);
        encode.int32(self.id(self.params_id));
        encode.int32(transfer_id);
        ROOT.get_bytes(&KEY_CALL).set_value(&encode.data());

        unsafe {
            if self.results_id != ptr::null_mut() {
                *self.results_id = get_object_id(1, KEY_RETURN, TYPE_MAP);
            }
        }
    }

    pub fn of_contract(&self, h_contract: ScHname) -> ScView {
        let mut ret = self.clone();
        ret.h_contract = h_contract;
        ret
    }

    fn id(&self, params_id: *mut i32) -> i32 {
        unsafe {
            if params_id == ptr::null_mut() {
                return 0;
            }
            *params_id
        }
    }
}

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

#[derive(Clone, Copy)]
pub struct ScFunc {
    delay: i32,
    h_contract: ScHname,
    h_function: ScHname,
    params_id: *mut i32,
    results_id: *mut i32,
    transfer_id: i32,
}

impl ScFunc {
    pub fn zero() -> ScFunc {
        ScFunc {
            delay: 0,
            h_contract: ScHname(0),
            h_function: ScHname(0),
            params_id: ptr::null_mut(),
            results_id: ptr::null_mut(),
            transfer_id: 0,
        }
    }

    pub fn new(h_contract: ScHname, h_function: ScHname, params_id: *mut i32, results_id: *mut i32) -> ScFunc {
        unsafe {
            if params_id != ptr::null_mut() {
                *params_id = ScMutableMap::new().map_id();
            }
        }

        ScFunc {
            delay: 0,
            h_contract: h_contract,
            h_function: h_function,
            params_id: params_id,
            results_id: results_id,
            transfer_id: 0,
        }
    }

    pub fn call(&self) {
        self.call_with_transfer(self.transfer_id);
    }

    fn call_with_transfer(&self, transfer_id: i32) {
        let mut encode = BytesEncoder::new();
        encode.hname(&self.h_contract);
        encode.hname(&self.h_function);
        encode.int32(self.id(self.params_id));
        encode.int32(transfer_id);
        ROOT.get_bytes(&KEY_CALL).set_value(&encode.data());

        unsafe {
            if self.results_id != ptr::null_mut() {
                *self.results_id = get_object_id(1, KEY_RETURN, TYPE_MAP);
            }
        }
    }

    pub fn delay(&self, seconds: i32) -> ScFunc {
        let mut ret = self.clone();
        ret.delay = seconds;
        ret
    }

    pub fn of_contract(&self, h_contract: ScHname) -> ScFunc {
        let mut ret = self.clone();
        ret.h_contract = h_contract;
        ret
    }

    pub fn post(&self) {
        self.post_to_chain(ROOT.get_chain_id(&KEY_CHAIN_ID).value())
    }

    pub fn post_to_chain(&self, chain_id: ScChainId) {
        if self.transfer_id == 0 {
            panic("transfer is required for post")
        }

        let mut encode = BytesEncoder::new();
        encode.chain_id(&chain_id);
        encode.hname(&self.h_contract);
        encode.hname(&self.h_function);
        encode.int32(self.id(self.params_id));
        encode.int32(self.transfer_id);
        encode.int32(self.delay);
        ROOT.get_bytes(&KEY_POST).set_value(&encode.data());
    }

    fn id(&self, params_id: *mut i32) -> i32 {
        unsafe {
            if params_id == ptr::null_mut() {
                return 0;
            }
            *params_id
        }
    }

    pub fn transfer(&self, transfer: ScTransfers) -> ScFunc {
        let mut ret = self.clone();
        ret.transfer_id = transfer.transfers.obj_id;
        ret
    }

    pub fn transfer_iotas(&self, amount: i64) -> ScFunc {
        self.transfer(ScTransfers::iotas(amount))
    }
}
